### DISCUSSION TITLE
Support for Global EMM Import Specification (GEMMIS)

### DISCUSSION THREAD

**Discussion by murkyness** [dosemu2/fdpp#173] [General]
Hi! Really happy to have found this project. 

Does fdpp currently, or will it in the future, support the GEMMIS API? 

This API is required for full MS-DOS feature parity, as well as running Windows 3.1 in 386 Enhanced mode. Adding it would certainly help make fdpp live up to its name as an enhanced version of FreeDOS.

**Comment by stsp** [comment]
Its not a DOS business to support
GEMMIS API. It is supported by emm386.exe.
I test windows-3.1 under fdpp regularly
and it works fine. But GEMMIS is entirely
unrelated to this and to fdpp.

> Adding it would certainly help make fdpp live up to its name as an enhanced version of FreeDOS.

FreeDOS is self-hosted, and fdpp needs
some run-time environment. So its not a
direct competitor to freedos, except for
the environments that can boot fdpp. And
on such environments fdpp is already the
unbeatable, much enhanced version of
FreeDOS. Unfortunately, such environments
are rare. :) GEMMIS will not help with that.

**Reply by murkyness** [reply]
Could I just take EMM386.EXE from a copy of MS-DOS and add it to fdpp?

**Reply by stsp** [reply]
What for?
GEMMIS is not the only bootstrap
protocol of win31. With the lack of
one, the only difference is that win31
would start in 286 mode by default,
but you can still use `win /3` to force.
As for win95 - you can't.

**Reply by murkyness** [reply]
That's interesting, since FreeDOS does not support `win /3`. Does FDPP really support 386-Enhanced Windows even though FreeDOS doesn't? 

I'm pretty sure that without GEMMIS, Windows 3.11 will only run in 286 mode, and anything above WfW 3.11 will fail to run.

**Reply by stsp** [reply]
> I'm pretty sure that without GEMMIS, Windows 3.11 will only run in 286 mode

OK, I'll stop that.
Please be sure in anything you want,
instead of spending 2 minutes to try
things out.
Windows 3.11 doesn't even have 286
mode at all, 3.1 is the last one that had it.

**Reply by jschwartzenberg** [reply]
> That's interesting, since FreeDOS does not support `win /3`.

AFAIK both FreeDOS and FDPP can both run `win /3` fine when you use them in DOSEMU2. I've also read that it would not be supported in pure FreeDOS, but in DOSEMU2 `win /3` works fine in most DOSs.

If you search around the dosemu2 organization you'll find issues that are still open pertaining exclusively to Windows 3.1 in 386 mode. For instance https://github.com/dosemu2/dosemu2/issues/381. Most software runs fine however. You can also find what is likely preventing Windows 9x from running in DOSEMU2, but you have to look at bit deeper in the tickets, https://github.com/dosemu2/dosemu2/issues/988.

I still need to write documentation on drivers to use, but I think this should be a good starting point:
https://github.com/jschwartzenberg/dos-download-scripts/blob/master/win31drv

In the olden days there used to be a project called Win4Lin that could run all Windows 9x versions in a virtual DOS environment on GNU/Linux, with patches to both kernels.

**Reply by stsp** [reply]
@jschwartzenberg are you subscribed to
an ML? You can try to help there as well:
https://marc.info/?l=linux-msdos&r=1&b=202103&w=2

**Reply by jschwartzenberg** [reply]
Nope, I just signed up. I didn't realize that mailing list was still in use, I think I was there in a long ago past :)

**Reply by stsp** [reply]
Not much in the use.

**Reply by murkyness** [reply]
> > I'm pretty sure that without GEMMIS, Windows 3.11 will only run in 286 mode
> 
> OK, I'll stop that.
> Please be sure in anything you want,
> instead of spending 2 minutes to try
> things out.
> Windows 3.11 doesn't even have 286
> mode at all, 3.1 is the last one that had it.

You're thinking of WFW 3.11

Also be nice :)

**Reply by murkyness** [reply]
> > That's interesting, since FreeDOS does not support `win /3`.
> 
> AFAIK both FreeDOS and FDPP can both run `win /3` fine when you use them in DOSEMU2. I've also read that it would not be supported in pure FreeDOS, but in DOSEMU2 `win /3` works fine in most DOSs.
> 
> If you search around the dosemu2 organization you'll find issues that are still open pertaining exclusively to Windows 3.1 in 386 mode. For instance [dosemu2/dosemu2#381](https://github.com/dosemu2/dosemu2/issues/381). Most software runs fine however. You can also find what is likely preventing Windows 9x from running in DOSEMU2, but you have to look at bit deeper in the tickets, [dosemu2/dosemu2#988](https://github.com/dosemu2/dosemu2/issues/988).
> 
> I still need to write documentation on drivers to use, but I think this should be a good starting point:
> https://github.com/jschwartzenberg/dos-download-scripts/blob/master/win31drv
> 
> In the olden days there used to be a project called Win4Lin that could run all Windows 9x versions in a virtual DOS environment on GNU/Linux, with patches to both kernels.

Thanks. Yes I was referring to pure freedos.

Do you know of a guide somewhere for installing Windows 3.11 on fdpp in 386 mode?

**Reply by stsp** [reply]
Yes, never seen 3.11 that
is not WfW. Hope it will work
on dosemu2 quite fine. But its
not tried.

> Also be nice :)

Sorry, I can't be nice.
You come and tell that dosbox
is more interesting for you w/o
trying either one.
Now you ask the guide to install
win3.11 instead of just running
its setup.exe.
I am pretty sure you will find 1024+
more reasons not to try anything out.

**Reply by murkyness** [reply]
I have been trying it out for a while... I'm still trying to get Windows 3.11 installed with dosemu2 as we speak. It's not working though. Why do you think I haven't tried either one?

Also, Dosbox-x is capable of running windows 98 stably, which AFAIK dosemu2 cannot. Observing this doesn't mean I was criticising dosemu2. I'm trying to use DOS programs on a stoney ridge laptop so speed is important, and in this case I think dosemu2 is a better fit than dosbox-x.

And yes, Windows 3.11 and WFW are different. You can get boot disks here: https://winworldpc.com/product/windows-3/311

**Reply by stsp** [reply]
> Why do you think I haven't tried either one?

Because then you would say something
like "I tried to install windows, but it crashes
for me this way" instead of talking about
GEMMIS and other unrelated things.

**Reply by murkyness** [reply]
> Because then you would say something
> like "I tried to install windows, but it crashes
> for me this way" instead of talking about
> GEMMIS and other unrelated things.

I extracted all the Windows 3.11 floppies to ~/.dosemu2/c_drive/INSTALL 
Then I started dosemu2 and ran SETUP from within the INSTALL directory

Windows setup starts 

![image](https://user-images.githubusercontent.com/70567689/114109538-b4267b00-9892-11eb-8d5d-aa144a9ebff1.png)

But when I try to continue it gets stuck at this screen

![image](https://user-images.githubusercontent.com/70567689/114109599-cef8ef80-9892-11eb-9136-940adf79ab4d.png)

and Windows never starts.

I'm assuming I need a custom config file but I haven't found that documented anywhere.

If you could help, I would appreciate it.

**Reply by stsp** [reply]
You can fill a bug report.
When you do, you will be prompted
automatically with what info to provide.

**Reply by murkyness** [reply]
> You can fill a bug report.
> When you do, you will be prompted
> automatically with what info to provide.

See, that's why I didn't bring it up here. lol

**Reply by stsp** [reply]
Hi again.

So it seems, @jschwartzenberg have
filled the bugs for you and also did the
bisect, so now we seem to have all the
win31 problems fixed.
Please try your tests again if you are
still interested.
While still no DOS prompt support in
386 mode, the DOS prompt of 286
mode is now much more reliable and
can Alt-Tab from the prot mode DOS
apps.

**Comment by stsp** [comment]
You might be interested in this project:
https://fossbytes.com/developer-creates-dos-subsystem-for-linux-dsl/
If they support GEMMIS, they could
bootstrap out of emm386.exe.

**Comment by murkyness** [comment]
I see. Thank you for the explanation.

I'm trying to run Windows 98SE, or at least WFW 3.11 in 386 Enhanced mode, on newer embedded hardware. I was interested your project since it runs on linux (allowing me to use linux drivers) and runs faster than DOSBox. Although the DSL link you provided is interesting, it looks like the opposite of what I'm trying to do.

I think DOSBox (or DOSBox-x) is closer to my use case, so I'll look more into that. But thank you again for your help :)

**Reply by stsp** [reply]
Hi @murkyness 
Just wanted to let you know that
we recently closed the remaining
windows-related instabilities with KVM.
Needs a rather recent linux kernel
though, as the problem was in the
kernel.

So while I can't promise much, at
least now the windows support that
we have, works at a full power.
Whereas when you tested it first, it
was completely unmaintained.
While there is still no dos box in
386 mode, the 286 mode dos box
works at full, including the Alt-Tab
from protected-mode apps - something
that on real PC is only possible in
386 mode.
Would be nice to know if your
experience with win31 under fdpp
is improved by these changes.

**Reply by stsp** [reply]
https://www.youtube.com/watch?v=QeM6trlaNL4

Someone made a nice video
on installing win31 under dosemu2.

**Reply by jschwartzenberg** [reply]
Interesting also that it uses an ET4000 display driver. I never knew that could work for DOSEMU2.
